### ISSUE TITLE
typename argument in Cursor.var()

### ISSUE THREAD

**Issue by doerwalter** [oracle/python-cx_Oracle#231]
I would like to be able to pass a value `None` for the `typename` argument of `Cursor.var()`. However `cx_Oracle` raises the following exception in this case:

```
DatabaseError: DPI-1046: parameter name cannot be a NULL pointer
```

I would have expected this to behave exactly like when `typename` wasn't passed at all.

This means that currenly I have to do the following:

```python
if typename is None:
    var = cursor.var(t)
else:
    var = cursor.var(t, typename=typename)
```

which is inelegant.

**Comment by doerwalter** [comment]
I would be even better if I could pass the `cx_Oracle` object type as the `type` argument to `Cursor.var()`, i.e. something that `Connection.gettype()` returns.

**Comment by anthony-tuininga** [comment]
Thanks @doerwalter. I agree that None should be allowable for the typename argument, and that the object returned from Connection.gettype() should be allowed to be passed to the dataType parameter of Cursor.var(). I'll address these both and let you know when they have been done.

**Comment by anthony-tuininga** [comment]
@doerwalter, I have just pushed changes that address both of these issues. Can you give them a whirl and let me know if the changes address your requests? Thanks!

**Comment by doerwalter** [comment]
Both changes work indead as expected. Thanks for implementing those. With this change I can then simply use

```python
arg.type if isinstance(arg, cx_Oracle.Object) else type(arg)
```

as the cursor type for passing the argument `var`.

But this leads to an even simpler solution (at least from the standpoint of the user ;)). Maybe what `Connection.gettype()` returns should be a real Python type, then the value could be an instance of that type. I.e. `Connection.gettype()` would return a *subclass* of `cx_Oracle.Object` and the value would be an instance of that subclass. Then the variable type would simply be `type(arg)` in all cases and `cx_Oracle.ObjectType` would be unneccessary. Conceptually this would be much simpler, because then types and instances would be much closer to how they work in Python. However I don't know whether that's feasable to implement.

**Comment by anthony-tuininga** [comment]
That's an interesting concept, which I considered briefly at one point. :-) I'll give it some more thought as I agree that it would simplify matters considerably and I believe it is indeed feasible to implement. I'll get back to you on that.

**Comment by anthony-tuininga** [comment]
The initial request was implemented in cx_Oracle 7.1 which was released yesterday. I still haven't looked into your other suggestion so I'll leave this open until I do.

**Comment by anthony-tuininga** [comment]
I've looked in this further and in Python 3.x this is definitely possible. Assuming I have the time available, I'll add it into cx_Oracle 8.x. I'll create a new enhancement request, however, so that this one can be closed.

**Comment by anthony-tuininga** [comment]
@doerwalter, I have made a number of changes in development which are at least somewhat related to this. Specifically, I have modifed ``var.type`` so that it is either (a) a database type constant or (b) an object type. In addition, all of the variable types are now the same class. What that means is that ``type(var(int)) == type(var(str))``. I considered creating a separate attribute -- so that ``var.type`` remains solely for object types and a new attribute like ``var.dbtype`` would be used for the type constant -- but thought this would be simpler. If you have any feedback, let me know! Thanks.
